File: diff_drive_example/src/diff_drive_example.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <std_msgs/msg/float64.hpp>

namespace diff_drive_example {

  class DiffDriveExample : public rclcpp::Node {
    public:
      DiffDriveExample()
      : rclcpp::Node("diff_drive_example")
      {
        sub_twist_ = create_subscription<geometry_msgs::msg::Twist>("cmd_vel", 1,
            std::bind(&DiffDriveExample::recv_twist, this, std::placeholders::_1));

        pub_left_ = create_publisher<std_msgs::msg::Float64>("left_speed", 1);
        pub_right_ = create_publisher<std_msgs::msg::Float64>("right_speed", 1);
      }

    private:
      rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr sub_twist_;
      rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr pub_left_;
      rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr pub_right_;

      void recv_twist(const geometry_msgs::msg::Twist::ConstSharedPtr msg) {
        double rw = 0.3;
        double W = 1.2;

        double v = msg->linear.x;
        double pdot = msg->angular.z;

        double left_speed = (v - 0.5 * W * pdot) / rw;
        double right_speed = (v + 0.5 * W * pdot) / rw;

        std_msgs::msg::Float64 left_speed_msg;
        std_msgs::msg::Float64 right_speed_msg;

        left_speed_msg.data = left_speed;
        right_speed_msg.data = right_speed;

        pub_left_->publish(left_speed_msg);
        pub_right_->publish(right_speed_msg);
      }

  };

}

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);
  auto node_instance = std::make_shared<diff_drive_example::DiffDriveExample>();
  rclcpp::spin(node_instance);
  rclcpp::shutdown();
  return 0;
}

File: diff_drive_example/scripts/diff_drive_example.py
#! /usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from std_msgs.msg import Float64


class DiffDriveExample(Node):
    def __init__(self):
        super().__init__('diff_drive_example')
        self.sub_twist = self.create_subscription(topic='cmd_vel', msg_type=Twist, callback=self.recv_twist, qos_profile=1)
        self.pub_left = self.create_publisher(topic='left_speed', msg_type=Float64, qos_profile=1)
        self.pub_right = self.create_publisher(topic='right_speed', msg_type=Float64, qos_profile=1)

    def recv_twist(self, msg: Twist):
        rw = 0.3
        W = 1.2

        v = msg.linear.x
        pdot = msg.angular.z

        left_speed = (v - 0.5 * W * pdot) / rw
        right_speed = (v + 0.5 * W * pdot) / rw

        left_speed_msg = Float64(data=left_speed)
        right_speed_msg = Float64(data=right_speed)

        self.pub_left.publish(left_speed_msg)
        self.pub_right.publish(right_speed_msg)


if __name__ == '__main__':
    rclpy.init()
    node_instance = DiffDriveExample()
    rclpy.spin(node_instance)
